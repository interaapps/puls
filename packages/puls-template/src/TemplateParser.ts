import {ParserAttribute, ParserOutput, ParserTag} from "./parser-types";
import {TemplateParserBase} from "./TemplateParserBase";

export class TemplateParser extends TemplateParserBase {
    autoCloseTags = ['hr', 'br', 'input', 'img', 'meta', 'link', 'wbr', 'source', 'keygen', 'spacer', 'isindex', 'track', 'param', 'embed', 'base', 'area', 'col', '!doctype']

    readTag(): ParserTag|null {
        let tag: ParserTag = { type: 'element', tag: '', attributes: [], body: [], from: this.index, to: 0 }
        let opened = false
        let tagNameOpened = true
        let closingTag = false
        let i = 0
        while (this.hasNext()) {
            const index = i++

            const {type, value} = this.get()

            if (!opened) {
                if (tagNameOpened) {
                    if (this.isWhiteSpace(value)) {
                        tag.attributes = this.readAttributes()

                        if (this.get().value === '>') {
                            tagNameOpened = false
                            continue
                        }

                        if (this.nextIs('/>')) {
                            this.next(2)
                            break
                        }
                    } else if (value === '>') {
                        tagNameOpened = false
                        continue
                    } else if (type === 'value') {
                        tag.tag = value
                    } else if (value !== '<') {
                        tag.tag += value
                    }
                }


                if (value === '>') {
                    if (typeof tag.tag === 'string' && this.autoCloseTags.includes(tag.tag.toLowerCase())) {
                        this.next()
                        break
                    }
                    opened = true
                }
            } else if (closingTag) {
                break
            } else {
                tag.body = this.readContent(['script', 'style'].includes(tag.tag as string), tag)
                closingTag = true
                this.next()
                break
            }

            this.next()
        }

        tag.to = this.index

        if (this.filterElements && !this.filterElements(tag)) {
            return null
        }

        return tag
    }

    isClosingTag(tag: ParserTag, ind = -1) {
        const index = this.index - ind
        const next = `</`

        // TODO: Check why -1 is needed
        if (this.nextIs(next, ind)) {
            let isEnding = false
            const afterClosingSlash = this.get(index + 2)

            if (afterClosingSlash.type === 'value') {
                if (afterClosingSlash.value === tag.tag) {
                    isEnding = true
                    tag.attributes.push(['slot', tag.body])
                    this.next()
                }
            } else {
                if (typeof tag.tag === 'string' && !this.nextIs(tag.tag)) {
                    isEnding = true
                    this.next(tag.tag.length)
                }
            }

            if (isEnding) {
                this.next(next.length - 1)
                this.skipEmpty()
                // Skipping >
                this.next()
            }

            return isEnding
        }
        return false
    }

    readAttributes(): ParserAttribute[] {
        const attributes: ParserAttribute[] = []
        let inAttribute = false

        while (this.hasNext()) {
            this.skipEmpty()
            const {type, value} = this.get()

            if (value === '>' || value === '/') {
                break;
            } else {
                const {name, value: attrValue} = this.readAttribute()
                attributes.push([name, attrValue])

                if (this.get().value === '>' || (this.get().value === '/' && this.get(this.index + 1)?.value === '>')) {
                    break;
                }
            }

            this.next()
        }
        return attributes
    }

    readAttribute(): {name: string, value: any, isLast: boolean, from: number, to: number} {
        const out = {name: '', value: true, isLast: false, from: this.index, to: 0}

        let hasReadName = false
        let hasReadEquals = false
        let foundValue = false

        while (this.hasNext()) {
            this.skipEmpty()
            const {type, value} = this.get()

            if (value === '>')
                break

            if (!hasReadName) {
                let doBreak = false

                if (type === 'value') {
                    out.name = value
                    hasReadName = true
                    continue
                }

                let [{value: name}] = this.readUntil(({type, value}) => {
                    if (value === '>') {
                        doBreak = true
                        out.isLast = true
                        return true
                    }
                    if (value === '=' || this.isWhiteSpace(value)) {
                        if (value === '=') {
                            return true
                        }

                        let i = this.index
                        while (this.get(i) && this.isWhiteSpace(this.get(i).value)) {
                            i++
                        }
                        if (this.get(i).value !== '=') {
                            doBreak = true
                            return true
                        }
                    }
                })

                if (name === '>') break

                if (!out.name)
                    out.name = name.trim()
                if (doBreak) break
                hasReadName = true
                continue
            } else if (!hasReadEquals) {
                if (value !== '=' || value === '>') {
                    break
                }
                hasReadEquals = true
            } else {
                if (type === 'value') {
                    out.value = value
                    this.next()
                    break
                } else if (!foundValue) {
                    let opened = false
                    let val = ''
                    let opener = ''
                    let i = 0

                    let [{value}] = this.readUntil(({type, value}) => {
                        let doNotAdd = false

                        if (i++ === 0) {

                            // attribute=${...} with inner spacing and quotes
                            if (this.nextIs('${')) {
                                this.next()

                                val += value + this.readUntilEndingBrace()
                                doNotAdd = true
                            }
                            if (value === '"' || value === "'")
                                opener = value
                            else
                                opener = ' '
                        }

                        if (opener === ' ' && (this.isWhiteSpace(value) || value === '>' || this.nextIs('/>'))) {
                            return true
                        }

                        if (opener !== ' ' && value === opener) {
                            if (opened) {
                                return true
                            }
                            opened = true
                        }
                        if (!doNotAdd)
                            val += value
                    })

                    out.value = opener === ' ' ? value : value.substring(1)
                    break
                }
            }

            this.next()
        }

        out.to = this.index
        return out
    }

    shipComment() {
        if (this.nextIs('<!--')) {
            while (this.hasNext()) {
                if (this.nextIs('-->')) {
                    this.next(3)
                    break;
                }
                this.next()
            }
            return true
        }
        return false
    }

    readContent(ignoreTags = false, tag: ParserTag|false = false): ParserOutput[] {
        const contents: ParserOutput[] = []
        let currentString = ''
        this.skipEmpty()
        while (this.hasNext()) {
            const { type, value } = this.get()

            if (type === 'char') {
                if (value === '<' && !this.isWhiteSpace(this.get(this.index + 1)?.value)) {
                    const currentIndex = this.index
                    if (tag !== false && this.isClosingTag(tag, 0)) {
                        break
                    } else {
                        this.index = currentIndex
                        if (ignoreTags) {
                            currentString += '<'
                            this.next()
                            continue
                        }
                    }

                    if (currentString !== '') {
                        contents.push({type: 'text', value: currentString})
                        currentString = ''
                    }

                    if (this.shipComment()) continue

                    const newTag = this.readTag()
                    if (newTag !== null)
                        contents.push(newTag)
                    this.skipEmpty()
                    continue
                }

            } else {
                if (currentString !== '') {
                    contents.push({type: 'text', value: currentString})
                    currentString = ''
                }
                contents.push({type, value})
                this.next()
                continue
            }
            currentString += value

            this.next()
        }

        if (currentString !== '') {
            contents.push({type: 'text', value: currentString})
        }

        return contents
    }
}