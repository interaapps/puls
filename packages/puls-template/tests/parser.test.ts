import TemplateParser from "../src/TemplateParser";

test('parser-text', () => {
    expect((TemplateParser.fromTemplate`Hello`).parse()).toMatchSnapshot()
})

test('parser-h1-element', () => {
    expect(TemplateParser.fromTemplate`<h1>Hello World!</h1>`.parse()).toMatchSnapshot()
})

test('parser-attributes', () => {
    expect(TemplateParser.fromTemplate`<h1 id="title" class="Hello World">Hello World!</h1>`.parse()).toMatchSnapshot()
})

test('parser-full-html', () => {
    expect(TemplateParser.fromTemplate`
        <!doctype html>
        <html lang="en">
        <head>
        <meta charset="UTF-8">
                     <meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
                                 <meta http-equiv="X-UA-Compatible" content="ie=edge">
                     <title>Document</title>
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
    </html>`.parse()).toMatchSnapshot()
})

test('parser-interpolation-attributes', () => {
    const a = () => alert('test')
    expect(TemplateParser.fromTemplate`
        <div id=${'hello-world'}>I'm a div</div>
    `.parse()).toMatchSnapshot()
})

test('parser-function-attributes', () => {
    const a = () => alert('test')
    expect(TemplateParser.fromTemplate`
        <button @click=${a}>Click me</button>
    `.parse()).toMatchSnapshot()
})