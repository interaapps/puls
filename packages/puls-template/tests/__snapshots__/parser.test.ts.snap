// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser-attributes 1`] = `
[
  {
    "attributes": [
      [
        "id",
        "title",
      ],
      [
        "class",
        "Hello World",
      ],
    ],
    "body": [
      {
        "type": "text",
        "value": "Hello World!",
      },
    ],
    "from": 0,
    "tag": "h1",
    "to": 52,
    "type": "element",
  },
]
`;

exports[`parser-full-html 1`] = `
[
  {
    "attributes": [
      [
        "html",
        true,
      ],
    ],
    "body": [],
    "from": 9,
    "tag": "!doctype",
    "to": 24,
    "type": "element",
  },
  {
    "attributes": [
      [
        "lang",
        "en",
      ],
    ],
    "body": [
      {
        "attributes": [],
        "body": [
          {
            "attributes": [
              [
                "charset",
                "UTF-8",
              ],
            ],
            "body": [],
            "from": 73,
            "tag": "meta",
            "to": 95,
            "type": "element",
          },
          {
            "attributes": [
              [
                "name",
                "viewport",
              ],
              [
                "content",
                "width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0",
              ],
            ],
            "body": [],
            "from": 117,
            "tag": "meta",
            "to": 243,
            "type": "element",
          },
          {
            "attributes": [
              [
                "http-equiv",
                "X-UA-Compatible",
              ],
              [
                "content",
                "ie=edge",
              ],
            ],
            "body": [],
            "from": 277,
            "tag": "meta",
            "to": 330,
            "type": "element",
          },
          {
            "attributes": [],
            "body": [
              {
                "type": "text",
                "value": "Document",
              },
            ],
            "from": 352,
            "tag": "title",
            "to": 375,
            "type": "element",
          },
        ],
        "from": 58,
        "tag": "head",
        "to": 391,
        "type": "element",
      },
      {
        "attributes": [],
        "body": [
          {
            "attributes": [],
            "body": [
              {
                "type": "text",
                "value": "Hello World!",
              },
            ],
            "from": 417,
            "tag": "h1",
            "to": 438,
            "type": "element",
          },
        ],
        "from": 400,
        "tag": "body",
        "to": 454,
        "type": "element",
      },
    ],
    "from": 33,
    "tag": "html",
    "to": 466,
    "type": "element",
  },
]
`;

exports[`parser-function-attributes 1`] = `
[
  {
    "attributes": [
      [
        "@click",
        [Function],
      ],
    ],
    "body": [
      {
        "type": "text",
        "value": "Click me",
      },
    ],
    "from": 9,
    "tag": "button",
    "to": 43,
    "type": "element",
  },
]
`;

exports[`parser-h1-element 1`] = `
[
  {
    "attributes": [],
    "body": [
      {
        "type": "text",
        "value": "Hello World!",
      },
    ],
    "from": 0,
    "tag": "h1",
    "to": 21,
    "type": "element",
  },
]
`;

exports[`parser-interpolation-attributes 1`] = `
[
  {
    "attributes": [
      [
        "id",
        "hello-world",
      ],
    ],
    "body": [
      {
        "type": "text",
        "value": "I'm a div",
      },
    ],
    "from": 9,
    "tag": "div",
    "to": 34,
    "type": "element",
  },
]
`;

exports[`parser-text 1`] = `
[
  {
    "type": "text",
    "value": "Hello",
  },
]
`;
